### PR TITLE
KAFKA-14412: Generalise over RocksDB WriteBatch

### PR DESCRIPTION
**Pull request by nicktelford**
The type hierarchy of RocksDB's `WriteBatch` looks like this:

```
        +---------------------+
        | WriteBatchInterface |
        +---------------------+
                   ^
                   |
        +---------------------+
        |  AbstractWriteBatch |
        +---------------------+
                   ^
                   |
        +----------+----------+
        |                     |
 +------------+    +---------------------+
 | WriteBatch |    | WriteBatchWithIndex |
 +------------+    +---------------------+
```

By switching our `BatchWritingStore` methods from `WriteBatch` to `WriteBatchInterface`, we enable the use of `WriteBatchWithIndex` as well.

### CODE REVIEW

**Comment by nicktelford** [comment]
This is part of KIP-892, and has been broken out into a separate PR to reduce the review burden on the main KIP-892 implementation, since it can be merged independently.

Note: KIP-892 makes use of `WriteBatchWithIndex` for its transaction buffer, so these changes are needed to enable us to use both `WriteBatch` (which continue to be used in `restoreBatch`) and `WriteBatchWithIndex` (which are used for transaction buffers).

@cadonna

**Comment by nicktelford** [comment]
@mjsax @ableegoldman @lucasbru @wcarlson5 @bbejeck @vvcephei @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: nit: log an error before we throw the exception

Also, imo this should just be an IllegalStateException instead of a ProcessorStateException

Comment (nicktelford) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Good point! Done.

**Comment by nicktelford** [comment]
I've rebased against trunk, so it's ready to be merged

**Review by ableegoldman — APPROVED** [review]
LGTM -- bonus points for the good-looking class hierarchy diagram in the PR description 😄

**Comment by ableegoldman** [comment]
Strangely one of the builds failed to compile due to some issue in the upgrade test/smoke client -- seems unrelated but we can't merge with a broken build, so let's retry and hope it goes away 🤞

**Comment by nicktelford** [comment]
There was a merge conflict with trunk that has now been resolved. @ableegoldman once the build passes, can we merge this?

**Comment by ableegoldman** [comment]
Two of the builds failed again...one of them looks like a failing Connect test bailed out with a sys exit, not sure what happened with the other one, but let's retry it...

**Comment by nicktelford** [comment]
I couldn't figure out how to restart the build, but I've rebased against trunk in the hope that some of these failing tests have been addressed upstream.

I'm confident they aren't introduced by this PR.

**Comment by ableegoldman** [comment]
You need to log in to restart the build, so it might be that only committers can trigger a retry...which is annoying (though I guess I kind of understand, don't want some random person to waste money & compute by spamming retries. We've had weirdos spam kafka before so it's not unthinkable 🤷‍♀️ )

Anyways I replayed it but looks like now there's some issue with Jenkins and/or git itself, so I replayed it again just now. FWIW I really don't think this PR is breaking anything, but I just can't merge with a completely broken build. Hopefully this time it goes through 🤞 
Screenshot 2023-12-14 at 12 41 17 PM

**Comment by nicktelford** [comment]
The build still appears to be failing, this time with an error from git about updating the HEAD... It doesnt look like other PRs are affected by this, so I'm guessing the build is cached per-PR and something has borked the cached build for this PR.

Shall I just close this PR and open a new one from the same branch? Might give it a fresh build that can then succeed.

**Comment by ableegoldman** [comment]
~~I honestly don't know...maybe try doing a rebase from trunk and force pushing again? That can often clear up this sort of thing. Ideally we don't have to abandon this PR for a fresh one (just to avoid losing history/review/comments), but that can be a last resort I suppose~~

~~Sorry for the trouble...if it's any consolation, this isn't the only PR that's been struggling with unrelated build failures of some kind or another lately. Although I'm guessing that's not actually any consolation to you~~

Wait, nevermind -- it looks like the latest build didn't have any actual build failures, only (unrelated) test failures. So I think we can merge this now

**Pull request merged**
